From 50b9cad8bb5fbc7e366610dac5b26ec74cb4cf40 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Wed, 03 Mar 2021 15:55:38 +0200
Subject: [PATCH] [SW-20154] check for safe limits for ppboot_phy bin

The ppboot + phy binary is read from flash and hence
need to be checked for safe operating values before
the DMA happens.

Change-Id: I5ddfc0d46d459cd52eba864ea57d71e250654331
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 9137d5c..1d7b9f8 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -291,6 +291,11 @@
 	}
 	arcboot_rom_status_update_bringup(image_progress, PROGRESS_BRINGUP_START);
 	flash_image_offset = arcboot_flash_image_offset(image_ptr);
+
+	/* check if offset is within safe limits */
+	if ((flash_image_offset + FLASH_64KB_PADD) >= FLASH_VALID_MASK)
+		return ARCBOOT_BRINGUP_ERROR;
+
 	flash_ppboot_pcie_offset = (uint32_t)(FLASH_64KB_PADD +
 					      flash_image_offset);
 	ppboot_pcie_size = (uint16_t)spi_flash_readl(flash_ppboot_pcie_offset);
@@ -299,11 +304,19 @@
 		return ARCBOOT_BRINGUP_ERROR;
 	}
 
+	/* check for min size as we are subtrating SECUREBOOT_IMG_HEAD_FOOTER */
+	if (ppboot_pcie_size <= SECUREBOOT_IMG_HEAD_FOOTER)
+		return ARCBOOT_BRINGUP_ERROR;
+
 	/* update in PSOC_SECURITY the current img fw_version */
 	flash_image_fw_ver = spi_flash_readl(flash_image_offset +
 					     FW_VERSION_SIZE);
 	security_report_img_fw(arcboot_rom_global.curr_img, flash_image_fw_ver);
 
+	/* check for safe limits */
+	if ((flash_ppboot_pcie_offset + ppboot_pcie_size) >= FLASH_VALID_MASK)
+		return ARCBOOT_BRINGUP_ERROR;
+
 	dma_copy_image(ppboot_pcie_size, flash_ppboot_pcie_offset, 0,
 		       SPI_DMA_DST_FW_SRAM);
 	if (dma_wait_finish() == DMA_CRC_ERROR) {
